types/sentry.d.ts

```diff
@@ -48,31 +48,35 @@ export interface TracingConfiguration {
     browserOptions?: Partial<BrowserTracing['options']>
 }
 
-export interface ModuleConfiguration {
-    clientConfig?: BrowserOptions
-    clientIntegrations?: IntegrationsConfiguration
-    config?: SentryOptions
-    customClientIntegrations?: string
-    customServerIntegrations?: string
-    disableClientRelease?: boolean
-    disableClientSide?: boolean
-    disabled?: boolean
-    disableServerRelease?: boolean
-    disableServerSide?: boolean
-    dsn?: string
-    tracing?: boolean | TracingConfiguration
-    initialize?: boolean
-    lazy?: boolean | LazyConfiguration
-    logMockCalls?: boolean
-    /** See available options at https://github.com/getsentry/sentry-webpack-plugin */
-    publishRelease?: boolean | Partial<SentryCliPluginOptions>
-    runtimeConfigKey?: string
-    serverConfig?: SentryOptions
-    serverIntegrations?: IntegrationsConfiguration
-    sourceMapStyle?: WebpackOptions.Devtool
-    requestHandlerConfig?: Handlers.RequestHandlerOptions
+type DeepPartial<T> = {
+  [P in keyof T]?: T[P] extends Array<infer I>
+    ? Array<DeepPartial<I>>
+    : DeepPartial<T[P]>;
 }
 
-interface ResolvedModuleConfiguration extends Omit<Required<ModuleConfiguration>, 'publishRelease'> {
-    publishRelease?: Partial<SentryCliPluginOptions>
+export interface ModuleConfiguration {
+  clientConfig: BrowserOptions
+  clientIntegrations: IntegrationsConfiguration
+  config: SentryOptions
+  customClientIntegrations: string
+  customServerIntegrations: string
+  disableClientRelease: boolean
+  disableClientSide: boolean
+  disabled: boolean
+  disableServerRelease: boolean
+  disableServerSide: boolean
+  dsn: string
+  tracing: boolean | TracingConfiguration
+  initialize: boolean
+  lazy: boolean | LazyConfiguration
+  logMockCalls: boolean
+  /** See available options at https://github.com/getsentry/sentry-webpack-plugin */
+  publishRelease: boolean | SentryCliPluginOptions
+  runtimeConfigKey: string
+  serverConfig: SentryOptions
+  serverIntegrations: IntegrationsConfiguration
+  sourceMapStyle: WebpackOptions['devtool']
+  requestHandlerConfig: Handlers.RequestHandlerOptions
 }
+
+export interface DeepPartialModuleConfiguration extends DeepPartial<ModuleConfiguration> {}
```

Comment:
Can you explain the use of `DeepPartial` here rather than just `Partial`? I would think that only the root properties should be made optional. Once the root property is provided then the values should be provided per original definition.

Unless there are some specific exception that I'm aware of.

Reply:
Well, it looks like it it's OK to do DeepPartial right now, at least.

Reply:
My idea was if someone were to use some nested options but not all the object like someone wants to add tracing but lower the sample rate 
`{
...
tracing: { tracesSampleRate: 0.4, },
...
}`. 
They would input something like this so the other values would be undefined or am I missing something?

Reply:
Yes, that's fine but I would expect in this case that `TracingOptions` properties are already specified as optional.

I'm thinking that it would not be correct in case we have an option like:

```ts
foo?: FooOptions
```

where `FooOptions` has some required properties like:

```ts
needsthis: string
bar?: string
```

In that case we'd want to preserve original types of `FooOptions`.

That said, I don't think there are currently any options that require that pattern but there seem to be options that need the opposite actuallly (like `publishRelease?: boolean | Partial<SentryCliPluginOptions>`).

Reply:
The reason I didn't go that way is that after we merge the user values with the default values we alwasy get the complete object. 

The only value I can see we might want to force the user to provide is the authToken if they want to publish the release, all the other values can be defaulted to the default values